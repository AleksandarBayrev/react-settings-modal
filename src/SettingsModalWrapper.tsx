import React from 'react';
import { IMessagePublisher } from './interfaces';
import { SettingsModal, SettingsModalState } from './SettingsModal';
import './styles.css';

export type SettingsModalWrapperProps = {
    messagePublisher: IMessagePublisher;
    state: SettingsModalState;
}

export type SettingsModalWrapperState = {
    isModalOpen: boolean;
}

export class SettingsModalWrapper extends React.Component<SettingsModalWrapperProps, SettingsModalWrapperState> {
    private settingsStorage: SettingsModalState
    public static readonly componentName: string = 'SettingsModalWrapper'
    constructor(props: SettingsModalWrapperProps) {
        super(props);
        this.state = {
            isModalOpen: false            
        };
        this.settingsStorage = this.props.state;
        this.props.messagePublisher.subscribe('settingsUpdated', async (state: SettingsModalState) => {
            this.settingsStorage = state;
        });
    }

    private openModal() {
        this.setState({
            isModalOpen: true
        });
    }

    private closeModal() {
        this.setState({
            isModalOpen: false
        });
    }

    render() {
        return (
            <div className='settings-modal-wrapper'>
                <div className='open-modal' onClick={() => this.openModal()}>
                    <span className='label'>Open modal</span>
                </div>
                <div className='close-modal label' onClick={() => this.closeModal()}>
                    <span className='label'>Close modal</span>
                </div>
                {this.state.isModalOpen ? <SettingsModal messagePublisher={this.props.messagePublisher} state={this.settingsStorage} /> : <></>}
            </div>
        )
    }
}