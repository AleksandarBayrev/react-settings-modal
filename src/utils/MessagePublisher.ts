import { IMessagePublisher, MessagePublisherCallback } from "../interfaces";

export type MessagePublisherSubscriptions = {
    [key: string]: MessagePublisherCallback[];
}

export class MessagePublisher implements IMessagePublisher {
    private readonly subscriptions: MessagePublisherSubscriptions
    constructor() {
        this.subscriptions = {};
    }

    /**
     * Used to call the subscribed callbacks to the topic asynchronously.
     * @param topic the topic on which we subscribe and publish messages to
     * @param data the data to be used in the callbacks
     * @returns Promise with empty string
     */
    async publish(topic: string, data?: any): Promise<string> {
        if (!this.subscriptions[topic]) {
            return Promise.reject(`No subscribers for ${topic} available`);
        }
        await Promise.all(this.subscriptions[topic].map((x) => x(data)));
        return Promise.resolve('');
    }

    /**
     * Used to subscribe a given topic with a callback.
     * @param topic the topic on which we subscribe and publish messages to
     * @param callback the callback that will be added to the topic array
     * @returns Promise with void value
     */
    subscribe(topic: string, callback: MessagePublisherCallback): Promise<void> {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (!this.subscriptions[topic]) {
                    this.subscriptions[topic] = [
                        callback
                    ];
                    return resolve();
                }
                this.subscriptions[topic].push(callback);
                return resolve();
            })
        });
    }

    /**
     * Unsubscribes the last callback subscribed to the topic 
     * @throws an error when no subscribers for the given topic are available
     * @param topic the topic on which we subscribe and publish messages to
     * @returns Promise with void value
     */
    unsubscribeOne(topic: string): Promise<void> {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (!this.subscriptions[topic]) {
                    return reject(`No subscribers for ${topic} available`);
                }
                this.subscriptions[topic].pop();
                return resolve();
            });
        });
    }

    /**
     * Unsubscribes all callbacks subscribed to the topic 
     * @throws an error when no subscribers for the given topic are available
     * @param topic the topic on which we subscribe and publish messages to
     * @returns Promise with void value
     */
    unsubscribeAll(topic: string): Promise<void> {
        return new Promise((resolve, reject) => {
            setTimeout(() => {
                if (!this.subscriptions[topic]) {
                    return reject(`No subscribers for ${topic} available`);
                }
                while (this.subscriptions[topic].length !== 0) {
                    this.subscriptions[topic].pop();
                }
                return resolve();
            });
        });
    }

}