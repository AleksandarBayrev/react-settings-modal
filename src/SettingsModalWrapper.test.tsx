import React from 'react';
import renderer from 'react-test-renderer';
import { SettingsModalWrapper } from './SettingsModalWrapper';
import { MessagePublisher } from './utils/MessagePublisher';

describe('SettingsModalWrapper', () => {
    it('does not render SettingsModal', () => {
        const component = renderer.create(<SettingsModalWrapper
            messagePublisher={new MessagePublisher}
            state={{
                isAPIEnabled: false
        }} />).toJSON();
        expect(component).toMatchSnapshot();
    });
    it('renders SettingsModal on clicking openModal', () => {
        const component = renderer.create(<SettingsModalWrapper
            messagePublisher={new MessagePublisher}
            state={{
                isAPIEnabled: false
        }} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
})