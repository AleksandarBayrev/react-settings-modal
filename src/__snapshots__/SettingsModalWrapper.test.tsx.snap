// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsModalWrapper does not render SettingsModal 1`] = `
<div
  className="settings-modal-wrapper"
>
  <div
    className="open-modal"
    onClick={[Function]}
  >
    <span
      className="label"
    >
      Open modal
    </span>
  </div>
  <div
    className="close-modal label"
    onClick={[Function]}
  >
    <span
      className="label"
    >
      Close modal
    </span>
  </div>
</div>
`;

exports[`SettingsModalWrapper renders SettingsModal on clicking openModal 1`] = `
<div
  className="settings-modal-wrapper"
>
  <div
    className="open-modal"
    onClick={[Function]}
  >
    <span
      className="label"
    >
      Open modal
    </span>
  </div>
  <div
    className="close-modal label"
    onClick={[Function]}
  >
    <span
      className="label"
    >
      Close modal
    </span>
  </div>
</div>
`;
